Use minified version of React

gaearon opened the pull request:
To the best of my knowledge websites should be minified before deploying to production.
In fact I learned this from Google web developer resources (thank you!)

This seems like an essential missing step in a benchmark that, if I understand it correctly, includes network response time. Minifying is especially important for React because it uses Uglify's dead-code elimination to strip out DEV-only code branches and comments.

JS payload before this change: 577 kB (120 kB gzipped)

<img width="371" alt="screen shot 2016-12-31 at 02 22 00" src="https://cloud.githubusercontent.com/assets/810438/21575300/d9acaba8-ceff-11e6-91e7-93edb90167f2.png">


JS payload after this change: 277 kB (48 kB gzipped)

<img width="473" alt="screen shot 2016-12-31 at 14 05 24" src="https://cloud.githubusercontent.com/assets/810438/21577850/16449e42-cf62-11e6-81e4-81852af5ed13.png">

So this is more than 50% reduction.

<s>There are alternative, more popular ways to achieve this, such as by adding Uglify. However it currently chokes on ES6 syntax and I don't know if adding an ES2015 Babel preset is something you want to do here. Vendoring React files is probably a good idea anyway (e.g. faster build times) but I couldn't figure out how to make Rollup do it. For the reference, Rollup is not very popular in React ecosystem. Webpack and Browserify can easily do this but I don't know how to do the same with Rollup or if it supports vendoring. So I just resorted to using globals instead.</s>

I added Uglify and ES2015 preset (which Uglify currently requires). After doing this I noticed Rollup Uglify integration [produces a lot of junk code](https://github.com/GoogleChrome/sample-framework-boot/pull/10#issuecomment-269864301) so I changed the project to use Webpack. This is not an "expert" level change as we specifically recommend Webpack/Browserify on [React installation page](https://facebook.github.io/react/docs/installation.html#development-and-production-versions) and offer these exact instructions (use `DefinePlugin` and `UglifyJsPlugin`).

I hope that this makes the benchmark a little more realistic! Thanks for providing an opportunity to contribute to it.

Fixes #8.

gaearon commented:
Signed the CLA.

gaearon commented on src/react/ssr.html:
```diff
@@ -6038,6 +6038,8 @@ <h2 class="post__comment-author">@username_301 wrote</h2>
 
+  <script src="react.min.js"></script>
+  <script src="react-dom.min.js"></script>
   <script src="app.js" async></script>
```
This one I'm not sure about. Can I make them all `async` but guarantee order?
If not I guess we'll have to bundle and add a minification step (and thus add Babel ES2015 transform).

JakeChampion replied:
`async` does not guarantee order unfortunately. You can use `defer` though :-)

`defer` tells browser to download scripts but not to parse them until it has finished parsing the document, then to execute each script in the order seen.

marvinhagemeister commented:
@gaearon For rollup ES6->ES5 transformation there is [`rollup-plugin-babel`](https://github.com/rollup/rollup-plugin-babel) or [`rollup-plugin-buble`](https://gitlab.com/Rich-Harris/rollup-plugin-buble). Both are really easy to set up.

For babel, simply add a `.babelrc` as usual and tell rollup to use the babel plugin.

```js
// rollup.config.js
import babel from 'rollup-plugin-babel';

export default {
  // ...
  plugins: [
    babel()
  ]
}
```

gaearon commented:
I also just using Uglify + ES2015 but the bundle is still larger (334 kB). It seems like Rollup generates a ton of junk code in CommonJS mode:

<img width="825" alt="screen shot 2016-12-31 at 13 08 30" src="https://cloud.githubusercontent.com/assets/810438/21577642/20f16b0c-cf5a-11e6-999b-6c7d3889a802.png">

I wonder why Rollup was chosen for this demo.

marvinhagemeister commented:
@gaearon indeed, I get the same results as you. Perhaps @Rich-Harris can help out here.

gaearon commented:
It kinda feels like Uglify runs on modules but I want it to run on Rollup result instead.

marvinhagemeister commented:
Yeah, it seems like uglify isn't able to rewrite nearly most of the function names or variables. I must admit, I'm quite surprised by this.

gaearon commented:
Pushed another commit.


I added Uglify and ES2015 preset (which Uglify currently requires). After doing this I noticed Rollup Uglify integration [produces a lot of junk code](https://github.com/GoogleChrome/sample-framework-boot/pull/10#issuecomment-269864301) so I changed the project to use Webpack. This is not an "expert" level change as we specifically recommend Webpack/Browserify on [React installation page](https://facebook.github.io/react/docs/installation.html#development-and-production-versions) and offer these exact instructions (use `DefinePlugin` and `UglifyJsPlugin`).

paullewis commented:
I owe an apology. With being away at the end of the year and the past couple of weeks back I've simply not gotten to this. I will take the weekend with my kids as planned, and then on Monday I will try and get this merged and re-run the numbers.

I'm sorry it appeared that this PR wasn't important. It is. I'll get there.

Rich-Harris commented:
Looks like there were a couple of problems with the Rollup config:

* No `format` option was specified, meaning it defaulted to creating an ES module build. (Were there any exports or external imports it would have failed at runtime, but there aren't.) This meant that variables were *effectively* being declared in the global scope, which is why Uglify didn't mangle those names. (In newer versions of Rollup it will complain if you omit `format`, rather than silently defaulting to `format: 'es'`, to prevent this exact problem.)
* rollup-plugin-replace was being run *after* the commonjs transformation. That meant that it was including modules that are only needed in dev mode.

With both of those things fixed, the Rollup build is **47.2kb** zipped, compared to **48.6kb** as of this branch (2.9% smaller), and has a slight advantage in evaluation time. So if the goal is to create the leanest bundle possible, keeping Rollup but fixing the config would be the right move — though as @gaearon points out, if the goal is to align with create-react-app then switching to Webpack would make sense.

I'll prepare an alternative PR so you can decide which course of action you want to take.

hzoo commented on src/react/webpack.config.js:
```diff
@@ -17,27 +20,30 @@ const intro = `/*!
+      exclude: /node_modules/,
+      loader: 'babel-loader',
+      query: {
+        presets: ['latest'],
```
babel-preset-env acts the same as preset-latest when you don't use options if you want to use that.

and usage is similar the same: `presets: ['env']`. Going to add an uglify option later in https://github.com/babel/babel-preset-env/issues/134 otherwise can try babili.

paullewis commented:
Thanks - really great to get all the feedback here :)
